src/auth/src/user_provider/watch_file_user_provider.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::path::Path;
+use std::sync::mpsc::channel;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use async_trait::async_trait;
+use common_telemetry::info;
+use notify::{EventKind, RecursiveMode, Watcher};
+use notify_debouncer_full::{new_debouncer, DebounceEventResult};
+use snafu::ResultExt;
+
+use crate::error::{FileWatchSnafu, Result};
+use crate::user_provider::{authenticate_with_credential, load_credential_from_file};
+use crate::{Identity, Password, UserInfoRef, UserProvider};
+
+pub(crate) const WATCH_FILE_USER_PROVIDER: &str = "watch_file_user_provider";
+
+pub(crate) struct WatchFileUserProvider {
+    users: Arc<Mutex<HashMap<String, Vec<u8>>>>,
+}
+
+impl WatchFileUserProvider {
+    pub fn new(filepath: &str) -> Result<Self> {
+        let credential = load_credential_from_file(filepath)?;
+        let users = Arc::new(Mutex::new(credential));
+        let this = WatchFileUserProvider {
+            users: users.clone(),
+        };
+
+        let (tx, rx) = channel::<DebounceEventResult>();
+
+        let mut debouncer =
+            new_debouncer(Duration::from_secs(1), None, tx).context(FileWatchSnafu)?;
+        debouncer
+            .watcher()
+            .watch(Path::new(filepath), RecursiveMode::NonRecursive)
+            .context(FileWatchSnafu)?;
+        let filepath = filepath.to_string();
+        std::thread::spawn(move || {
+            let _hold = debouncer;
+            while let Ok(res) = rx.recv() {
+                if let Ok(events) = res {
+                    if events
+                        .iter()
+                        .any(|e| matches!(e.kind, EventKind::Modify(_) | EventKind::Create(_)))
```

Comment:
Do we need to process the `Delete` event? If the user-provider file is deleted, clear the user info.

Reply:
Here is an issue: What does it mean if the file is empty? Does it mean no one can be authed, or everyone can be authed?

I ever thought the later so that it's suitable for debugging, but it may frustrate users want to ensure no "whitelist for all".

You may find our code contains:

```
    ensure!(
        !credential.is_empty(),
        InvalidConfigSnafu {
            value: filepath.to_string(),
            msg: "UserProvider's file must contains at least one valid credential",
        }
    );
```

and that is why this is subtle.

As a developer, I like empty to no auth so that debugging between authed and noauthed is easy. But I'm not quite sure if it's desired as a feature.

Reply:
I think empty content is an invalid configuration, so the `ensure` is necessary and correct.
Empty files and deleting files are totally different, I think the file does not exist means noauth looks like making sense.

Reply:
Make sense. Let me apply all your comments above.